core/base/dim.hpp

```diff
@@ -0,0 +1,236 @@
+        : first_{size}, rest_{size}
+    {}
+
+    /**
+     * Creates a dimension object with the specified dimensions.
+     *
+     * If the number of dimensions given is less than the dimensionality of the
+     * object, the remaining dimensions are set to the same value as the last
+     * value given.
+     *
+     * For example, in the context of matrices `dim<2>{2, 3}` creates the
+     * dimensions for a 2-by-3 matrix.
+     *
+     * @param first  first dimension
+     * @param rest  other dimensions
+     */
+    template <typename... Rest>
+    constexpr GKO_ATTRIBUTES dim(const dimension_type &first,
+                                 const Rest &... rest)
+        : first_{first}, rest_{static_cast<dimension_type>(rest)...}
+    {}
+
+    /**
+     * Returns the requested dimension.
+     *
+     * For example, if `d`is a dim<2> object representing matrix dimensions,
+     * `d[0]` returns the number of rows, and `d[1]` returns the number of
+     * columns.
+     *
+     * @param dimension  the requested dimension
+     *
+     * @return the `dimension`-th dimension
+     */
+    constexpr GKO_ATTRIBUTES const dimension_type &operator[](
+        const size_type &dimension) const noexcept
+    {
+        return GKO_ASSERT(dimension < dimensionality), *(&first_ + dimension);
+    }
+
+    /**
+     * @copydoc operator[]() const
+     */
+    GKO_ATTRIBUTES dimension_type &operator[](
+        const size_type &dimension) noexcept
+    {
+        return GKO_ASSERT(dimension < dimensionality), *(&first_ + dimension);
+    }
+
+    /**
+     * Checks if all dimensions evaluate to true.
+     *
+     * For standard arithmetic types, this is equivalent to all dimensions being
+     * different than zero.
+     *
+     * @return true if and only if all dimensions evaluate to true
+     */
+    constexpr GKO_ATTRIBUTES operator bool() const
+    {
+        return static_cast<bool>(first_) && static_cast<bool>(rest_);
+    }
+
+    /**
+     * Checks if two dim objects are equal.
+     *
+     * @param x  first object
+     * @param y  second object
+     *
+     * @return true if and only if all dimensions of both objects are equal.
+     */
+    friend constexpr GKO_ATTRIBUTES bool operator==(const dim &x, const dim &y)
+    {
+        return x.first_ == y.first_ && x.rest_ == y.rest_;
+    }
+
+    /**
+     * Multiplies two dim objects.
+     *
+     * @param x  first object
+     * @param y  second object
+     *
+     * @return a dim object representing the size of the tensor product `x * y`
+     */
+    friend constexpr GKO_ATTRIBUTES dim operator*(const dim &x, const dim &y)
+    {
+        return dim(x.first_ * y.first_, x.rest_ * y.rest_);
+    }
+
+private:
+    constexpr GKO_ATTRIBUTES dim(const dimension_type first,
+                                 dim<dimensionality - 1> rest)
+        : first_{first}, rest_{rest}
+    {}
+
+    dimension_type first_;
+    dim<dimensionality - 1, dimension_type> rest_;
```

Comment:
Just a question so that I can better understand C++, why not have a static array of dimension_types and do recursive classes instead? I would say having a static array is more efficient, at least for compile time (if not also for `operator*`). In terms of code, I don't think there is a need for a base case specialization also.

I mean something like that:
```C++
#include <cstddef>

template<std::size_t Dimensionality, typename DimensionType= std::size_t>
class dim{
public:
    template <typename... Rest>
    constexpr dim(const DimensionType &first,
                  const Rest &... rest)
    : dims{first, static_cast<DimensionType>(rest)...}
    {}

private:
    DimensionType dims[Dimensionality];
};

void foo()
{
    dim<3> dim{1, 4, 3};
}
```

Reply:
Try implementing `operator ==`, `operator *` as `constexpr` functions with your approach.
Also, the constructor in your example is not equivalent to the one I implemented, `dim<3>(3)` creates dimensions (3, 3, 3) in my code, and `(3, 0, 0)` in your.

I don't know how I would do that with your definition, but if you find an elegant way, we can sure drop the recursive definition.